src/services/elasticsearch/models/users.js

```diff
@@ -0,0 +1,130 @@
+import { app } from '../../../index.js'
+import sql from 'sql-template-tag'
+
+const index = 'users'
+const indices = {
+  index,
+  fields: {
+    //Full text search
+    first_name: { type: 'text' },
+    last_name: { type: 'text' },
+    username: { type: 'keyword' },
+    email: { type: 'keyword' },
+    created_at: { type: 'date' },
+
+    //Filters
+    causes_tags: { type: 'keyword' }, //Filter: Social Causes
+    skills: { type: 'keyword' }, //Filter: Skills
+    city: { type: 'keyword' }, //Filter: Location
+    country: { type: 'keyword' }, //Filter: Location
+    timezone: { type: 'keyword' }, //Filter: Timezone
+    // experience_level: { type: 'keyword' },//Filter: Experience Level //TODO: add to object after adding to preferences on users?
+    // payment_type: { type: 'keyword' }, //Filter: Payment Type //TODO: add to object after adding to preferences on users?
+    /*Payment Scheme: ? //TODO: add to object after adding to preferences on users?
+        > Fixed: Multiple options
+        > Hourly: Range
+    */
+    open_to_volunteer: { type: 'keyword' }, //Filter: Open To Volunteer
+    preferences: {
+      properties: {
+        title: { type: 'keyword' },
+        value: { type: 'keyword' },
+        title_value: { type: 'keyword' }
+      }
+    }
+  }
+}
+
+function transformer(document) {
+  return {
+    id: document.id,
+    first_name: document.first_name,
+    last_name: document.last_name,
+    username: document.username,
+    email: document.email,
+    created_at: document.created_at,
+    causes_tags: document.causes_tags ?? [],
+    skills: document.skills ?? [],
+    city: document.city,
+    country: document.country,
+    timezone: document.timezone,
+    open_to_volunteer: document.open_to_volunteer,
+    preferences: document.preferences.map((preference) => {
+      return {
+        ...preference,
+        title_value: `${preference.title}:${preference.value}`
+      }
+    })
+  }
+}
+
+const indexing = async ({ id }) => {
+  const user = await app.db.get(
+    sql`
+    SELECT *, gn.timezone as timezone,
+    COALESCE(
+      jsonb_agg(
+        json_build_object(
+          'title', p.title, 
+          'value', p.value
+        ) 
+      ) FILTER (WHERE p.id IS NOT NULL),
+      '[]'
+    ) as preferences
+    FROM users u
+    LEFT JOIN geonames gn ON gn.id=u.geoname_id
+    LEFT JOIN preferences p ON p.identity_id=u.id
+    WHERE u.id=${id}
+    GROUP BY u.id, gn.timezone
+    `
+  )
+
+  const document = transformer(user)
+
+  const indexingDocuments = [app.searchClient.indexDocument(index, document.id, document)]
+  try {
+    return await Promise.all(indexingDocuments)
+  } catch (e) {
+    console.log(e)
+  }
+}
+
+const initIndexing = async () => {
+  const { rows } = await app.db.query(
```

Comment:
this wont work you query all records to ram ? you should paginate data from db to ram then do bulk indexing

Reply:
> this wont work you query all records to ram ? you should paginate data from db to ram then do bulk indexing

✅ Provided chunk-based ( or pagination-based syncing ), solved a lot of performance issues ( either database and node.js runtime memory itself ) Please confirm if i followed the right approach ( i will apply them to users and orgs also if you confirm this )